File: apps/mobile/modules/habitflow-widget/src/HabitflowWidgetModule.ts
import { NativeModule, requireOptionalNativeModule } from "expo";

/**
 * The native surface of the widget mirror. Android-only, and optional even
 * there: Expo Go and the web build have no native module, and `setMirror`
 * must degrade to a no-op rather than throw from inside a mutation.
 */
declare class HabitflowWidgetModule extends NativeModule {
    /** Replace the mirror payload. Null clears it (sign-out, deletion). */
    setMirror(json: string | null): Promise<void>;
    /** Whether the launcher currently has any HabitFlow widget placed. */
    hasWidgets(): boolean;
}

export default requireOptionalNativeModule<HabitflowWidgetModule>(
    "HabitflowWidget",
);

File: apps/mobile/modules/habitflow-widget/android/src/main/java/com/habitflow/widget/HabitflowWidgetModule.kt
package com.habitflow.widget

import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition

/**
 * The JS → native half of the widget mirror. Everything it does is a single
 * SharedPreferences write plus a redraw broadcast, so it is safe to call from
 * any mutation path without worrying about cost.
 */
class HabitflowWidgetModule : Module() {
    override fun definition() = ModuleDefinition {
        Name("HabitflowWidget")

        /**
         * Replace the mirror with `json`, then redraw every placed widget.
         * Passing null clears it, which is what sign-out and account deletion
         * do: a widget still showing a deleted account's habits is a privacy
         * bug, not a stale cache.
         */
        AsyncFunction("setMirror") { json: String? ->
            WidgetMirror.write(appContext.reactContext ?: return@AsyncFunction, json)
        }

        /** Whether the launcher currently has any HabitFlow widget placed. */
        Function("hasWidgets") {
            val context = appContext.reactContext ?: return@Function false
            WidgetSizes.anyPlaced(context)
        }
    }
}
